src/components/views/rooms/MessageComposerInput.js

```diff
@@ -425,164 +564,313 @@ export default class MessageComposerInput extends React.Component {
-                let editorRoot = this.refs.editor.refs.editor.parentNode.parentNode;
-                editorRoot.scrollTop = editorRoot.scrollHeight;
-            }
-        });
+    richToMdEditorState(editorState: Value): Value {
+        // FIXME: this conversion loses pills (turning them into pure MD links).
+        // We need to add a pill-aware deserialize method
+        // to PlainWithPillsSerializer which recognises pills in raw MD and turns them into pills.
+        return Plain.deserialize(
+            // FIXME: we compile the MD out of the RTE state using slate-md-serializer
+            // which doesn't roundtrip symmetrically with commonmark, which we use for
+            // compiling MD out of the MD editor state above.
+            this.md.serialize(editorState),
+            { defaultBlock: DEFAULT_NODE }
+        );
     }
 
     enableRichtext(enabled: boolean) {
-        if (enabled === this.state.isRichtextEnabled) return;
+        if (enabled === this.state.isRichTextEnabled) return;
 
-        let contentState = null;
+        let editorState = null;
         if (enabled) {
-            const md = new Markdown(this.state.editorState.getCurrentContent().getPlainText());
-            contentState = RichText.htmlToContentState(md.toHTML());
+            editorState = this.mdToRichEditorState(this.state.editorState);
         } else {
-            let markdown = RichText.stateToMarkdown(this.state.editorState.getCurrentContent());
-            if (markdown[markdown.length - 1] === '\n') {
-                markdown = markdown.substring(0, markdown.length - 1); // stateToMarkdown tacks on an extra newline (?!?)
-            }
-            contentState = ContentState.createFromText(markdown);
+            editorState = this.richToMdEditorState(this.state.editorState);
         }
 
         Analytics.setRichtextMode(enabled);
 
         this.setState({
-            editorState: this.createEditorState(enabled, contentState),
-            isRichtextEnabled: enabled,
+            editorState: this.createEditorState(enabled, editorState),
+            isRichTextEnabled: enabled,
+        }, ()=>{
+            this.refs.editor.focus();
         });
+
         SettingsStore.setValue("MessageComposerInput.isRichTextEnabled", null, SettingLevel.ACCOUNT, enabled);
-    }
+    };
+
+    /**
+    * Check if the current selection has a mark with `type` in it.
+    *
+    * @param {String} type
+    * @return {Boolean}
+    */
+
+    hasMark = type => {
+        const { editorState } = this.state
+        return editorState.activeMarks.some(mark => mark.type == type)
+    };
+
+    /**
+    * Check if the any of the currently selected blocks are of `type`.
+    *
+    * @param {String} type
+    * @return {Boolean}
+    */
+
+    hasBlock = type => {
+        const { editorState } = this.state
+        return editorState.blocks.some(node => node.type == type)
+    };
+
+    onKeyDown = (ev: Event, change: Change, editor: Editor) => {
```

Comment:
s/Event/KeyboardEvent/

which made me realise `KeyboardEvent::keyCode` is depr.

Reply:
done (on your branch)